### PR TITLE
Enable c++20 CI and move CUDA CI to ubuntu 22.04

### LINKED ISSUE #592: Add C++20 build to CI
We should start testing DLA-Future with C++20 and in particular the P2300 reference implementation to make the transition to it smoother.

### PR DESCRIPTION
**Pull request by rasolca**
Fixes #592

### CODE REVIEW

File: spack/packages/dla-future/package.py [hunk 1]
```diff
@@ -18,6 +18,13 @@ class DlaFuture(CMakePackage, CudaPackage):
 
     version("develop", branch="master")
 
+    cxxstds = ('17', '20')
+    variant('cxxstd',
+            default='17',
+            values=cxxstds,
+            description='Use the specified C++ standard when building')
+    conflicts('cxxstd=20', when='+cuda')
```
[the remark this answers is not in the document]
Comment (msimberg) [reply spack/packages/dla-future/package.py:26]: https://github.com/eth-cscs/DLA-Future/issues/595

**Comment by rasolca** [comment]
bors try

**Comment by msimberg** [comment]
That looks very suspicious :/ Do I see correctly that the CUDA+clang setup used to be clang 10 and now you've updated it to clang 12? Is the CUDA version unchanged?

**Comment by rasolca** [comment]
I took the chance to update to ubuntu22.04 CUDA builds as well. Unfortunately only cuda 11.7.0 is available with this constraint.
So yeah both clang and cuda changed.
What really puzzle me is that gcc11 builds runs correctly (same base image).

**Comment by rasolca** [comment]
Let me try different clang version on ubuntu20.04 and ubuntu22.04. (All with CUDA 11.7)

bors try

**Comment by rasolca** [comment]
Summary: all the clang builds fail with CUDA 11.7. (Note clang14 build passed as spack fell back to gcc11.2)

**Comment by rasolca** [comment]
Try now with CUDA 11.6.2 and different clang versions

bors try

**Comment by rasolca** [comment]
Let me try CUDA 11.5.2 as well.

bors try

**Comment by rasolca** [comment]
Summary:
- Cuda 11.6 and 11.7 + clang10-13 -> zgemm segfault (reproduced removing pika and umpire dependencies from test_blas_tile, but not reproducible compiling a simple test in the container)
- Cuda 11.3 - 11.5 + clang 10 and 12 when possible -> compilation problem
- Cuda 11.1 and 11.2 + clang 10 works.

Reverting to Cuda 11.1 for clang build.

bors try

**Review by msimberg — APPROVED** [review]

**Pull request merged**